modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java

```diff
@@ -0,0 +1,177 @@
+    public static Collection<Object[]> params() {
+        Collection<Object[]> params = new ArrayList<>();
+
+        for (CacheAtomicityMode cacheMode: CacheAtomicityMode.values()) {
+            for (boolean updateMode: new boolean[] {true, false}) {
+                for (boolean touchMode: new boolean[] {true, false})
+                    params.add(new Object[] { updateMode, touchMode, cacheMode });
+            }
+        }
+
+        return params;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        if (!cfg.isClientMode()) {
+            cfg.setDataStorageConfiguration(new DataStorageConfiguration()
+                    .setDefaultDataRegionConfiguration(new DataRegionConfiguration().setPersistenceEnabled(true)))
+                .setCacheConfiguration(new CacheConfiguration<>(DEFAULT_CACHE_NAME)
+                    .setAtomicityMode(mode)
+                    .setBackups(1)
+                    .setExpiryPolicyFactory(TouchedExpiryPolicy.factoryOf(new Duration(TimeUnit.SECONDS, 10))));
+        }
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /** */
+    @Test
+    public void concurrentUpdateTouchCheckpointKey() throws Exception {
+        startGrids(2).cluster().state(ClusterState.ACTIVE);
+
+        IgniteEx cln = startClientGrid(2);
+
+        AtomicBoolean stop = new AtomicBoolean();
+
+        IgniteInternalFuture<?> updateFut = runMultiThreadedAsync(() -> {
+            while (!stop.get()) {
+                if (updateWithDataStreamer)
+                    stream(cln);
+                else
+                    cln.cache(DEFAULT_CACHE_NAME).put(KEY, rnd.nextInt());
+            }
+        }, 1, "update");
+
+        IgniteInternalFuture<?> touchFut = runMultiThreadedAsync(() -> {
+            while (!stop.get()) {
+                if (touchWithScanQuery)
+                    cln.cache(DEFAULT_CACHE_NAME).query(new ScanQuery<>()).getAll();
+                else
+                    cln.cache(DEFAULT_CACHE_NAME).get(KEY);
+            }
+
+        }, 1, "touch");
+
+        IgniteInternalFuture<?> cpFut = runMultiThreadedAsync(() -> {
+            while (!stop.get()) {
+                try {
+                    forceCheckpoint(F.asList(grid(0), grid(1)));
+                }
+                catch (Exception e) {
+                    // No-op.
+                }
+            }
+        }, 1, "checkpoint");
+
+        Thread.sleep(30_000);
```

Comment:
8 tests * 30 secs, 4 minutes total time. Perhaps too much, WDYT?

Reply:
On my machine it requires 10-20 secs to reproduce. Replace it with counter to 1000 checkpoint iterations.